src/backend/distributed/connection/remote_commands.c

```diff
@@ -184,12 +186,187 @@ LogRemoteCommand(MultiConnection *connection, const char *command)
+ * accepts a MultiConnection instead of a plain PGconn.  It makes sure it can
+ * send commands asynchronously without blocking (at the potential expense of
+ * an additional memory allocation).
  */
 int
 SendRemoteCommand(MultiConnection *connection, const char *command)
 {
+	PGconn *pgConn = connection->pgConn;
+	bool wasNonblocking = PQisnonblocking(pgConn);
+	int rc = 0;
+
 	LogRemoteCommand(connection, command);
-	return PQsendQuery(connection->pgConn, command);
+
+	/* make sure not to block anywhere */
+	if (!wasNonblocking)
+	{
+		PQsetnonblocking(pgConn, true);
+	}
+
+	rc = PQsendQuery(pgConn, command);
+
+	/* reset nonblocking connection to its original state */
+	if (!wasNonblocking)
+	{
+		PQsetnonblocking(pgConn, false);
+	}
+
+	return rc;
+}
+
+
+/*
+ * GetCommandResult is a wrapper around PQgetResult() that handles interrupts.
+ *
+ * If raiseInterrupts is true and an interrupt arrives, e.g. the query is
+ * being cancelled, CHECK_FOR_INTERRUPTS() will be called, which then throws
+ * an error.
+ *
+ * If raiseInterrupts is false and an interrupt arrives that'd otherwise raise
+ * an error, GetRemotecommandResult returns false.
+ *
+ * Handling of interrupts is important to allow queries being cancelled while
+ * waiting on remote nodes. In a distributed deadlock scenario cancelling
+ * might be the only way to resolve the deadlock.
+ */
+PGresult *
+GetRemoteCommandResult(MultiConnection *connection, bool raiseInterrupts)
+{
+	PGconn *pgConn = connection->pgConn;
+	int socket = 0;
+	int waitFlags = WL_POSTMASTER_DEATH | WL_LATCH_SET;
+	bool wasNonblocking = false;
+	PGresult *result = NULL;
+	bool failed = false;
+
+	/* short circuit tests around the more expensive parts of this routine */
+	if (!PQisBusy(pgConn))
+	{
+		return PQgetResult(connection->pgConn);
+	}
+
+	socket = PQsocket(pgConn);
+	wasNonblocking = PQisnonblocking(pgConn);
+
+	/* make sure not to block anywhere */
+	if (!wasNonblocking)
+	{
+		PQsetnonblocking(pgConn, true);
+	}
+
+	if (raiseInterrupts)
+	{
+		CHECK_FOR_INTERRUPTS();
+	}
+
+	/* make sure command has been sent out */
+	while (!failed)
+	{
+		int rc;
+
+		ResetLatch(MyLatch);
+
+		/* try to send all the data */
+		rc = PQflush(pgConn);
+
+		/* stop writing if all data has been sent, or there was none to send */
+		if (rc == 0)
+		{
+			break;
+		}
+
+		/* if sending failed, there's nothing more we can do */
+		if (rc == -1)
+		{
+			failed = true;
+			break;
+		}
+
+		/* this means we have to wait for data to go out */
+		Assert(rc == 1);
+
+		rc = WaitLatchOrSocket(MyLatch, waitFlags | WL_SOCKET_WRITEABLE, socket, 0);
+
+		if (rc & WL_POSTMASTER_DEATH)
+		{
+			ereport(ERROR, (errmsg("postmaster was shut down, exiting")));
+		}
+
+		if (rc & WL_LATCH_SET)
+		{
+			/* if allowed raise errors */
+			if (raiseInterrupts)
+			{
+				CHECK_FOR_INTERRUPTS();
+			}
+
+			/*
+			 * If raising errors allowed, or called within in a section with
+			 * interrupts held, return NULL instead.
```

Comment:
Yep, makes sense. We should be careful that there are a few cases where rollback on cancellation is currently undesirable (master_append_table_to_shard), but that should fall outside the scope of the coordinated transaction.